# Codon Optimization for Biopharming

In recent years, the technique of codon optimization has gained popularity in the biopharmaceutical industry to rapidly create and deploy high-yields of commercially valuable protein products. The tobbaco plant, *Nicotiana benthamiana*, is of particular interest due to the plant's rapid growth rate and ease of genetic manipulation. 

Currently, a range of external software tools exists to back-translate given protein sequences to DNA sequences with a user-specified codon usage table. However, the sole *N. benthamiana* [codon usage table currently existing](https://www.kazusa.or.jp/codon/cgi-bin/showcodon.cgi?species=4100) has not been updated in 15 years and is based on a limited number of genes (100 coding sequences). 

The objective of this project is to create a coding script to translate a given protein sequence into a DNA sequence based on a newly constructed codon use table for the target heterologous host, *N. benthamiana*, within 3 months.

## What Can I Do With This?

The code in this project can be used

### 1) A more efficient way to compare the codon usage of a set of proteins to existing/standard codon usages.

For example by 

### 2) Usage of the script to translate a given protein sequence into correspondign DNA sequence.

### 3) 

## Getting Started

These instructions will get you a copy of the project up and running on your local machine for development and testing purposes. See deployment for notes on how to deploy the project on a live system.

### Prerequisites/Dependencies

What things you need to install the software and how to install them

```
Give examples BLAST+ needs to be installed (what version?)
```

### Installation

A step by step series of examples that tell you how to get a development env running

Say what the step will be

```
Give the example
```

And repeat

```
until finished
```

End with an example of getting some data out of the system or using it for a little demo

## Example Usage

Explain how to run the automated tests for this system

### Example 1 - Using CDS Functions

Explain what these tests test and why

```
Give an example
```

### Example 2

Explain what these tests test and why

```
Give an example
```

## Authors

* **Anchitaa Ghag** - [anchitaaghag](https://github.com/anchitaaghag)

## Advisors

* **Dr. Andrew Hamilton-Wright** - [andrewhw](https://github.com/andrewhw) Associate Professor, Department of Computer Science, University of Guelph, Guelph, ON N1G 2W1, Canada. *(Computational/Informatics Expertise)*
* **Dr. Doug Cossar** - VP Research, PlantForm Corporation Canada, Toronto, ON M4S 3E2, Canada. *(External/Industry Advisor)*
* **Dr. Jennifer Geddes-McAlister** - Assistant Professor, Department of Molecular and Cellular Biology, University of Guelph, Guelph, ON N1G 2W1, Canada. *(Biological Expertise)*

## Acknowledgments

* This project is part of my major reasearch project (BINF 6999) to fullfill the requirements of the Master of Bioinformatics degree (University of Guelph, ON, Canada). Project Duration: May 2022 - August 2022. See also "Advisors" section above.
* Hat tip to anyone whose code was used
* Inspiration
* etc

## Template Credit

This file was created by adapting the README.md file template by Billie Thompson [PurpleBooth](https://github.com/PurpleBooth). The template can be found [here](https://gist.github.com/PurpleBooth/109311bb0361f32d87a2).
